opm/simulators/flow/BlackoilModelEbosNldd.hpp

```diff
@@ -0,0 +1,801 @@
+            // Otherwise hardcode to ILU0
+            if (domains_[index].cells.size() < 200) {
+                loc_param.linsolver_ = "umfpack";
+            } else {
+                loc_param.linsolver_ = "ilu0";
+                loc_param.linear_solver_reduction_ = 1e-2;
+            }
+            loc_param.linear_solver_print_json_definition_ = false;
+            domain_linsolvers_.emplace_back(model_.ebosSimulator(), loc_param);
+        }
+
+        assert(int(domains_.size()) == num_domains);
+    }
+
+    //! \brief Called before starting a time step.
+    void prepareStep()
+    {
+        // Setup domain->well mapping.
+        model_.wellModel().setupDomains(domains_);
+    }
+
+    //! \brief Do one non-linear NLDD iteration.
+    template <class NonlinearSolverType>
+    SimulatorReportSingle nonlinearIterationNldd(const int iteration,
+                                                 const SimulatorTimerInterface& timer,
+                                                 NonlinearSolverType& nonlinear_solver)
+    {
+        // -----------   Set up reports and timer   -----------
+        SimulatorReportSingle report;
+        Dune::Timer perfTimer;
+
+        model_.initialLinearization(report, iteration, nonlinear_solver.minIter(), timer);
+
+        if (report.converged) {
+            return report;
+        }
+
+        // -----------   If not converged, do an NLDD iteration   -----------
+
+        auto& solution = model_.ebosSimulator().model().solution(0);
+        auto initial_solution = solution;
+        auto locally_solved = initial_solution;
+
+        // -----------   Decide on an ordering for the domains   -----------
+        const auto domain_order = this->getSubdomainOrder();
+
+        // -----------   Solve each domain separately   -----------
+        std::vector<SimulatorReportSingle> domain_reports(domains_.size());
+        for (const int domain_index : domain_order) {
+            const auto& domain = domains_[domain_index];
+            SimulatorReportSingle local_report;
+            switch (model_.param().local_solve_approach_) {
+            case DomainSolveApproach::Jacobi:
+                solveDomainJacobi(solution, locally_solved, local_report,
+                                  iteration, timer, domain);
+                break;
+            default:
+            case DomainSolveApproach::GaussSeidel:
+                solveDomainGaussSeidel(solution, locally_solved, local_report,
+                                       iteration, timer, domain);
+                break;
+            }
+            // This should have updated the global matrix to be
+            // dR_i/du_j evaluated at new local solutions for
+            // i == j, at old solution for i != j.
+            if (!local_report.converged) {
+                // TODO: more proper treatment, including in parallel.
+                OpmLog::debug("Convergence failure in domain " + std::to_string(domain.index));
```

Comment:
You need `<string>` for `std::to_string()`.

Reply:
added.